magma/array.py

```diff
@@ -583,5 +549,367 @@ def unused(self):
+            new_value = o[i]
+            if curr_value is not new_value:
+                # Skip updating wire in the case that it's the same value
+                # (avoids an error message)
+                child.wire(new_value)
+
+    @debug_wire
+    def wire(self, o, debug_info):
+        o = magma_value(o)
+        self._check_wireable(o, debug_info)
+        if self._has_elaborated_children():
+            # Ensure the children maintain consistency with the bulk wire
+            self._wire_children(o)
+        else:
+            # Perform a bulk wire
+            Wireable.wire(self, o, debug_info)
+
+    def unwire(self, o):
+        if self._has_elaborated_children():
+            for i, child in self._enumerate_children():
+                child.unwire(o[i])
+        else:
+            Wireable.unwire(self, o)
+
+    def iswhole(self):
+        if self._has_elaborated_children():
+            return Array._iswhole(self._collect_children(lambda x: x))
+        return True
+
+    def const(self):
+        if self._has_elaborated_children():
+            return all(child.const()
+                       for _, child in self._enumerate_children())
+        return False
+
+    def _resolve_bulk_wire(self):
+        """
+        If a child reference is made, we "expand" a bulk wire into the
+        constiuent children to maintain consistency
+        """
+        if self._wire.driven():
+            # Remove bulk wire since children will now track the wiring
+            value = self._wire.value()
+            Wireable.unwire(self, value)
+
+            # Update children
+            for i in range(len(self)):
+                self._get_t(i).wire(value[i])
+
+    def _make_t(self, index):
+        if issubclass(self.T, MagmaProtocol):
+            return self.T._from_magma_value_(
+                self.T._to_magma_()(name=ArrayRef(self, index)))
+        else:
+            return self.T(name=ArrayRef(self, index))
+
+    def _resolve_slice_children(self, start, stop):
+        for i in range(start, stop):
+            # Avoid resolving each index more than once
+            if i not in self._ts:
+                self._get_t(i)
+
+    def _resolve_slice_driver(self, start, stop, value):
+        # When we encounter an overlapping slice that is already bulk driven,
+        # we ensure the corresponding children to their current values
+        if value._wire.driven():
+            driver = value._wire.value()
+            Wireable.unwire(value, driver)
+            for i in range(start, stop):
+                self._ts[i] @= driver[i - start]
+
+    def _remove_slice(self, key):
+        # Remove slice since we don't need to track it anymore (handled
+        # by child logic) to optimize other slice iteration logic
+        # This avoids having to iterate over slices when we are just
+        # using their children instead
+        if key in self._slices:
+            del self._slices[key]
+        if key[0] in self._slices_by_start_index:
```

Comment:
I think if we delete the slices ahead of time, and don't add when there's an overlap (since they'll be resolved and removed anyways), we can avoid the if check (and this gives us a sanity check where we'll get a key error if the assumption is violated)